### DISCUSSION TITLE
Bug preventing blob waveform rendering on iOS

### DISCUSSION THREAD

**Discussion by mulhoon** [katspaugh/wavesurfer.js#3466] [Q&A] [answered]
I'm not sure if this is a bug with wavesurfer.js per se, but it can certainly be fixed within wavesurfer.

Wavesurfer can't render a waveform from a blob loaded via the iOS file system (using capacitor, iOS WKWebView). This is only since version 7.4.7. But it can still play.

I'm loading the blob like this
```js
media: new Audio(file)
```

And I'm getting this error
```
Unhandled Promise Rejection: Error: Failed to fetch capacitor://localhost/_capacitor_file_/var/mobile/Containers/Data/Application/676246CD-2EA0-4...
```

The relevant code is here
[wavesurfer.js/src/fetcher.ts](https://github.com/katspaugh/wavesurfer.js/blob/19cd5544ac57af83e8a5843bfcbd806935babe2e/src/fetcher.ts#L43C1-L45C4)
```js
43  if (!response.ok) {
44    throw new Error(`Failed to fetch ${url}: ${response.status} (${response.statusText})`)
45  }
```

That's because on iOS, the response isn't well formed, but still has a body:

<img width="773" alt="image" src="https://github.com/katspaugh/wavesurfer.js/assets/1137909/15a6ecff-620e-4ba0-a8cf-c8ea5edb4ad1">
<br/><br/>
If I remove the response check, the waveform renders as normal.

## Possible solutions:
1. Don't throw an error at all 
2. Use the [bodyUsed](https://developer.mozilla.org/en-US/docs/Web/API/Request/bodyUsed) attribute instead. Needs further investigation.
3. Explicitly match 4xx and 5xx status codes

What do you think? I think we should probably not throw an error for now as it limits the use of this on iOS.

Many thanks
Nic

**Comment by katspaugh** [comment]
Oh, interesting. Thanks for investigating!

> Don't throw an error at all

That would be suboptimal because it can then fail further down the line with a confusing error.

> Use the [bodyUsed](https://developer.mozilla.org/en-US/docs/Web/API/Request/bodyUsed) attribute instead. Needs further investigation.

This attribute seems to be for the request body, not for the response, if I'm reading the MDN page right.

> Explicitly match 4xx and 5xx status codes

This could work. Something like `if (response.status >= 400)` perhaps? With a comment linking to this thread.

**Comment by mulhoon** [comment]
That sounds best. 
Do you want me to add it as a pull request?

**Reply by katspaugh** [reply]
Sure, that'd be awesome!

**Reply by mulhoon** [reply, marked as answer]
Fix added in pull request https://github.com/katspaugh/wavesurfer.js/pull/3470
